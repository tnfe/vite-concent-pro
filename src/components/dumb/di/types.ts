
export interface IDiCompPropsBase {
  block?: boolean,
  disabled?: boolean,
  required?: boolean,
  extraStyle?: Record<string, any>,
  remark?: string,
  placeholder?: string,
}