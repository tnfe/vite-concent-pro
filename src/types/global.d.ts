/**
 * 声明window里用户新增的属性，让ts编译通过
 */
interface SuperProfiler {
  new(): { container: any };
}

declare global {
  interface Window {
    _isPub: boolean;
    Raven: any;
    SuperProfiler: SuperProfiler;

    [key: string]: any;
  }
}
