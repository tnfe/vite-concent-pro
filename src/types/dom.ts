export type InputChangeEv =  React.ChangeEvent<HTMLInputElement>;

export type KeyboardEv =  React.KeyboardEvent<HTMLInputElement>;

export type MouseEv =  React.MouseEvent<HTMLInputElement>;
