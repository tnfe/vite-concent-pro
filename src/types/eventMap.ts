import * as ev from 'configs/constant/event';

/**
 * 管理所有事件名对应参数类型的 mappedType
 */
export interface EvMap {
  [ev.someEvent]: [number, string],
}
