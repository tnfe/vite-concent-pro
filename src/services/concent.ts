/**
 * concent 相关的一些公共封装函数
 */
import {
  useConcent, reducer, getState as getSt, getGlobalState as getGst, emit, getComputed,
  IActionCtxBase, cst, MODULE_DEFAULT, MODULE_GLOBAL,
  ICtxBase, IAnyObj, SettingsType, ComputedValType, SetupFn, MultiComputed, CallTargetParams,
} from 'concent';
import { noop } from 'utils/fn';
import { CtxM, CtxMConn, CtxConn, Modules, RootRd, RootState, RootCu } from 'types/store';
import { EvMap } from 'types/eventMap';

function priBuildCallParams(moduleName: Modules, connect: Array<Modules>, options?: Options<any, any, any, any, any, any>) {
  const targetOptions = options || {};
  const { setup, tag, extra, staticExtra, cuDesc, passCuDesc = true, props = {}, ccClassKey } = targetOptions;
  const regOpt = { module: moduleName, connect, setup, props, tag, extra, staticExtra, cuDesc: null as any };
  if (passCuDesc) regOpt.cuDesc = cuDesc;
  return { regOpt, ccClassKey };
}

/**
 * 调用目标函数，用于对接 reducer里的 ghost函数
 * @param callerParams
 * @param ac
 */
export async function callTarget<T extends any = any>(callerParams: CallTargetParams, ac: IActionCtxBase, errHandler?: (err: Error) => void): Promise<T> {
  let ret: T = {} as any;
  try {
    // 支持 reducer文件里内部调用 ac.dispatch(loading, fnPayload(targetFn, payload))
    if (Array.isArray(callerParams)) {
      ret = await Function.prototype.apply.call(ac.dispatch, ac, callerParams);
    } else {
      const { fnName, payload, renderKey, delay } = callerParams;
      ret = await ac.dispatch(fnName, payload, renderKey, delay) as unknown as any;
    }
  } catch (err) {
    if (errHandler) errHandler(err);
    throw err;
  }
  return ret;
}

export interface ValidSetup {
  (ctx: ICtxBase): IAnyObj | void;
}
export interface ValidMapProps {
  (ctx: ICtxBase): IAnyObj;
}
export interface OptionsBase<
  P extends IAnyObj,
  CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj,
  StaticExtra extends any,
  Mp extends ValidMapProps
  > {
  props?: P;
  tag?: string;
  ccClassKey?: string;
  extra?: Extra;
  staticExtra?: StaticExtra;
  /**
   * 一个遗留的参数，对接useConcent的mapProps，每一轮渲染都会调用，返回结果可通过 ctx.mapped拿到
   */
  mapProps?: Mp;
  cuDesc?: CuDesc;
  /**
   * 是否透传 cuSpec 给 useConcent函数，默认为true，
   * 表示需要透传，此时用户不需要再setup函数体内调用 ctx.computed(cuSpec)
   * 如果用户设置passCuSpec为false，表示传入 cuSpec 仅为了方便推导出refComputed类型，但不透传 cuSpec 给 useConcent函数
   * 注意此时用户需要在 setup函数体内调用 ctx.computed(cuSpec) 来完成示例计算函数的定义，
   * 否则 refComputed 里拿不到真正的计算结果
   */
  passCuDesc?: boolean;
}
export interface Options<
  P extends IAnyObj,
  Setup extends ValidSetup,
  CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj,
  StaticExtra extends any,
  Mp extends ValidMapProps
  > extends OptionsBase<P, CuDesc, Extra, StaticExtra, Mp> {
  setup?: Setup;
}

/**
 * 属于某个模块
 * use the target model context you want by passing a module name
 * 如需要全局任意地方可通过 useC2Mod('xx') 导出xx模块上下文来使用，
 * 需要在 src/models/index.js 显式的导出该模块
 *
 * -----------------------[Export example]-----------------------
 * ```js
 * import somePageModel from 'pages/SomePage/model';
 * import someCompModel from 'components/SomeComp/model';
 * const allModels = {...somePageModel,  ...someCompModel};
 * ```
 * ---------------------------------------------------------------
 *
 * 各种调用示例：
 * 需自己跳转到 components/_demosOfHook/useC2Mod 文件里查看
 * @see {useC2ModDemos.Example} - 简单示例
 * @see {useC2ModDemos.ExampleCallMr} - 调用moduleReducer
 * @see {useC2ModDemos.ExampleReadMcu} - 读取模块计算结果
 * @see {useC2ModDemos.ExampleSetup} - 配置setup
 * @see {useC2ModDemos.ExampleRefCu} - 定义实例计算
 * @param moduleName
 * @param options {Options} - 可选参数，见 Options定义
 */
export function useC2Mod<
  M extends Modules, Setup extends ValidSetup, P extends IAnyObj, CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj, StaticExtra extends any, Mp extends ValidMapProps,
  >(moduleName: M, options?: Options<P, Setup, CuDesc, Extra, StaticExtra, Mp>) {
  const { regOpt, ccClassKey } = priBuildCallParams(moduleName, [], options);
  type Ctx = CtxM<P, M, SettingsType<Setup>, ComputedValType<CuDesc>, [Extra, StaticExtra, ReturnType<Mp>]>;
  return useConcent<{}, Ctx>(regOpt, ccClassKey);
}

/**
 * 使用默认模块（在没有重写$$default模块的情况下，默认模块是一个空模块）
 * @param options
 * @returns
 */
export function useC2DefaultMod<
  Setup extends ValidSetup, P extends IAnyObj, CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj, StaticExtra extends any, Mp extends ValidMapProps,
  >(options?: Options<P, Setup, CuDesc, Extra, StaticExtra, Mp>) {
  const { regOpt, ccClassKey } = priBuildCallParams(cst.MODULE_DEFAULT, [], options);
  type Ctx = CtxM<P, MODULE_DEFAULT, SettingsType<Setup>, ComputedValType<CuDesc>, [Extra, StaticExtra, ReturnType<Mp>]>;
  return useConcent<{}, Ctx>(regOpt, ccClassKey);
}

/**
 * 属于某个模块，连接多个模块
 */
export function useC2ModConn<
  M extends Modules, Conn extends Array<Modules>, Setup extends ValidSetup, P extends IAnyObj, CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj, StaticExtra extends any, Mp extends ValidMapProps,
  >(moduleName: M, connect: Conn, options?: Options<P, Setup, CuDesc, Extra, StaticExtra, Mp>) {
  const { regOpt, ccClassKey } = priBuildCallParams(moduleName, connect, options);
  type Ctx = CtxMConn<P, M, Conn[number], SettingsType<Setup>, ComputedValType<CuDesc>, [Extra, StaticExtra, ReturnType<Mp>]>;
  return useConcent<{}, Ctx>(regOpt, ccClassKey);
}

/**
 * 连接多个模块
 * ```js
 *  const { connectedState, cr } = useC2Conn(['mod1', 'mod2']);
 *  // connectedState.mod1.stateXxx
 *  // cr.mod1.methodXxx
 *
 * ```
 */
export function useC2Conn<
  Conn extends Array<Modules>, Setup extends ValidSetup, P extends IAnyObj, CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj, StaticExtra extends any, Mp extends ValidMapProps,
  >(connect: Conn, options?: Options<P, Setup, CuDesc, Extra, StaticExtra, Mp>) {
  const { regOpt, ccClassKey } = priBuildCallParams(cst.MODULE_DEFAULT, connect, options);
  type Ctx = CtxConn<P, Conn[number], SettingsType<Setup>, ComputedValType<CuDesc>, [Extra, StaticExtra, ReturnType<Mp>]>;
  return useConcent<{}, Ctx>(regOpt, ccClassKey);
}

/**
 * 适用于不属于任何模块，只是设置setup函数的场景
 * @param setup
 * @param options - see OptionsBase
 */
export function useSetup<
  T extends SetupFn, P extends IAnyObj, CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj, StaticExtra extends any, Mp extends ValidMapProps,
  >(setup: T, options?: OptionsBase<P, CuDesc, Extra, StaticExtra, Mp>) {
  const mergedOptions = { setup, ...options };
  const { regOpt, ccClassKey } = priBuildCallParams(cst.MODULE_DEFAULT, [], mergedOptions);
  type Ctx = CtxM<P, MODULE_DEFAULT, SettingsType<T>, ComputedValType<CuDesc>, [Extra, StaticExtra, ReturnType<Mp>]>;
  const { settings } = useConcent<{}, Ctx>(regOpt, ccClassKey);
  return settings;
}

/**
 * useSetupCtx 会返回整个ctx对象，区别于 useSetup 只返回 settings
 * @param setup
 * @param options - see OptionsBase
 */
export function useSetupCtx<
  T extends SetupFn, P extends IAnyObj, CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj, StaticExtra extends any, Mp extends ValidMapProps,
  >(setup: T, options?: OptionsBase<P, CuDesc, Extra, StaticExtra, Mp>) {
  const mergedOptions = { setup, ...options };
  const { regOpt, ccClassKey } = priBuildCallParams(cst.MODULE_DEFAULT, [], mergedOptions);
  type Ctx = CtxM<P, MODULE_DEFAULT, SettingsType<T>, ComputedValType<CuDesc>, [Extra, StaticExtra, ReturnType<Mp>]>;
  const ctx = useConcent<{}, Ctx>(regOpt, ccClassKey);
  return ctx;
}


/**
 * 为属于某个模块的 ctx 标记类型, 通常使用在class成员变量上 和 setup函数体内
 * 在class组件成员变量使用时不需要传递第三位参数ctx，组件实例化时会由concent注入并替换
 * 在setup函数里使用时，可直接传入已经创建好的ctx
 */
export function typeCtxM<
  M extends Modules, Setup extends ValidSetup, P extends IAnyObj, CuDesc extends MultiComputed<any>,
  Extra extends IAnyObj, StaticExtra extends any, Mp extends ValidMapProps,
  >(moduleName: M, options?: Options<P, Setup, CuDesc, Extra, StaticExtra, Mp>, ctx?: any) {
  noop(moduleName, options);
  type Ctx = CtxM<P, M, SettingsType<Setup>, ComputedValType<CuDesc>, [Extra, StaticExtra, ReturnType<Mp>]>;
  return (ctx || {}) as Ctx;
}

/**
 * useC2Mod的工厂函数，返回钩子函数的同时，也提供了帮助推导setup函数的ctx参数类型的辅助函数
 * 注意! 此工厂函数仅适用于 setup函数ctx参数不需要感知 props, extra 类型时，方可使用
 * @param moduleName
 * @param options
 * -----------------------[Code example]-----------------------
 * ```js
 *  const ret = makeUseC2Mod("Counter");
 *   function setupA1(c: any) {
 *    const ctx = ret.typeCtx(c);
 *     const cu = ctx.computed({countX6: (n) => n.value * 6 });
 *     return { cu };
 *   }
 *   export function UseC2ModByFactory() {
 *     const ctx = ret.useC2Mod({ setup: setupA1 });
 *     return <h1>{ctx.state.bigValue} {ctx.settings.cu.countX6}</h1>
 *   }
 * ```
 * --------------------------------------------------------------
 */
export function makeUseC2Mod<M extends Modules>(moduleName: M) {
  return {
    /**
     * 需要传入的 setup 函数
     */
    useC2Mod: <Setup extends ValidSetup, P extends IAnyObj, CuDesc extends MultiComputed<any>,
      Extra extends IAnyObj, StaticExtra extends any, Mp extends ValidMapProps,
      >(options?: Options<P, Setup, CuDesc, Extra, StaticExtra, Mp>) => {
      const { regOpt, ccClassKey } = priBuildCallParams(moduleName, [], options);
      type Ctx = CtxM<P, M, SettingsType<Setup>, ComputedValType<CuDesc>, [Extra, StaticExtra, ReturnType<Mp>]>;
      return useConcent<P, Ctx>(regOpt, ccClassKey);
    },
    /**
     * 推导setup函数的ctx参数类型
     */
    typeCtx: (ctx: ICtxBase) => {
      return ctx as CtxM<{}, M>;
    },
  };
}

export const ccReducer = (reducer as unknown) as RootRd;

/**
 * 获取 global模块的状态
 * 在已拥有 concent model 上下文、action 上下文的地方，
 * 推荐直接获取，代替调用此函数，因为直接获取数据时组件并不会订阅数据变化
 */
export function getGlobalState() {
  const globalState = getGst<RootState>();
  return globalState;
}

/**
 * 获取整个根状态
 * 注意直接获取数据时组件并不会订阅数据变化
 */
export function getRootState() {
  const rootState = getSt<RootState>();
  return rootState;
}

/**
 * 获取目标模块状态
 * 注意直接获取数据时组件并不会订阅数据变化
 */
export function getModelState<T extends Modules>(modelName: T) {
  if (modelName) {
    const modelState = getSt<RootState, T>(modelName);
    return modelState;
  }
  throw new Error('miss modelName');
}

/**
 * 获取目标模块状态
 */
export function getModelComputed<T extends Modules>(modelName: T) {
  if (modelName) {
    const modelComputed = getComputed<RootCu, T>(modelName);
    return modelComputed;
  }
  throw new Error('miss modelName');
}

/**
 * 获取 global模块的状态
 * 在已拥有 concent model 上下文、action 上下文的地方，
 * 推荐直接获取，代替调用此函数，因为直接获取数据时组件并不会订阅数据变化
 */
export function getGlobalComputed() {
  const globalComputed = getComputed<RootCu, MODULE_GLOBAL>(cst.MODULE_GLOBAL);
  return globalComputed;
}

type EvKeys = keyof EvMap;
/**
 * 发射事件
 * @param eventName - 事件名
 * @param args
 */
export function ccEmit<E extends EvKeys, T extends EvMap[E]>(eventName: E, ...args: T) {
  emit(eventName, ...args);
}

/**
 * 携带id的发射事件
 * @param eventDesc - [eventName, id]
 * @param args
 */
export function ccEmitId<E extends EvKeys, T extends EvMap[E]>(eventDesc: [E, string], ...args: T) {
  emit(eventDesc, ...args);
}

type OnFn = <E extends EvKeys>(eventName: E, cb: (...args: EvMap[E]) => void) => void;
/**
 * 配合EvMap，为ctx.on装配类型
 * 外部调用时传入具体的事件名就推导出cb函数的参数列表类型
 *
    function setup(ctx: Ctx){
      const on = ctxOn(ctx);
      on('xxx',(a, b)=>{
       // 此处ts能感知a、b的具体类型
      })
    }
 */
export function ctxOn(ctx: ICtxBase) {
  return ctx.on as OnFn;
}

type OnIdFn = <E extends EvKeys>(eventDesc: [E, string], cb: (...args: EvMap[E]) => void) => void;
/**
 * 可以携带id的ctx.on
 * @param ctx
 */
export function ctxOnId(ctx: ICtxBase) {
  return ctx.on as OnIdFn;
}
