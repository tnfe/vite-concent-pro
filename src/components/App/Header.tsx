import React from 'react';
import { Layout, Popover, Switch, Tag, Avatar } from 'antd';
import { SwitchChangeEventHandler } from 'antd/lib/switch/index';
import { useConcent, SettingsType } from 'concent';
import { SketchPicker, ColorChangeHandler, ColorResult } from 'react-color'
import styles from './App.module.css';
import { SettingOutlined } from '@ant-design/icons';
import { CtxDe } from 'types/store';
import { Blank, VerticalBlank } from 'components/dumb/general';

interface ISettingPanelProps {
  color: string;
  onChangeComplete: ColorChangeHandler;
  onWebsiteColorChange: ColorChangeHandler;
  siderThemeChecked: boolean;
  onSiderThemeChange: SwitchChangeEventHandler;
  headerThemeChecked: boolean;
  onHeaderThemeChange: SwitchChangeEventHandler;
}

function SettingPanel({
  color, onChangeComplete, onWebsiteColorChange, siderThemeChecked, onSiderThemeChange, headerThemeChecked, onHeaderThemeChange,
}: ISettingPanelProps) {
  return (
    <div style={{ padding: '12px 28px' }}>
      <Tag color="geekblue">站点主题设置：</Tag>
      <VerticalBlank />
      <SketchPicker color={color} onChange={onWebsiteColorChange} onChangeComplete={onChangeComplete} />
      <VerticalBlank />
      <div>
        <Tag color="geekblue">边栏设置：</Tag>
        <Blank />
        <Switch checkedChildren="关闭暗黑边栏" unCheckedChildren="开启暗黑边栏" checked={siderThemeChecked} onChange={onSiderThemeChange} />
      </div>
      <VerticalBlank />
      <div>
        <Tag color="geekblue">顶栏设置：</Tag>
        <Blank />
        <Switch checkedChildren="关闭暗黑顶栏" unCheckedChildren="开启暗黑顶栏" checked={headerThemeChecked} onChange={onHeaderThemeChange} />
      </div>
    </div>
  );
}

type Se = SettingsType<typeof setup>;
function setup(ctx: CtxDe) {
  const { globalReducer: grd } = ctx;
  return {
    onWebsiteColorChange(colorResult: ColorResult) {
      grd.changeThemeColor(colorResult.hex);
    },
    onSiderThemeChange(checked: boolean) {
      grd.switchSiderTheme(checked);
    },
    onHeaderThemeChange(checked: boolean) {
      grd.switchHeaderTheme(checked);
    },
  }
}

function AppHeader() {
  const {
    globalState: st, globalComputed: gcu, settings: se
  } = useConcent<{}, CtxDe<{}, Se>>({ setup, tag: 'Header' });
  const uiContent = <SettingPanel color={st.themeColor} onWebsiteColorChange={se.onWebsiteColorChange}
    onChangeComplete={se.onWebsiteColorChange} siderThemeChecked={gcu.siderThemeSwitchChecked}
    headerThemeChecked={gcu.headerThemeSwitchChecked}
    onSiderThemeChange={se.onSiderThemeChange} onHeaderThemeChange={se.onHeaderThemeChange} />;

  return (
    <Layout.Header className={styles.header} style={gcu.headerStyle}>
      <div className={styles.userIconWrap}>
        <Avatar size={40} src={st.uesrIcon} />
        <Blank width="8px" />
        {st.userName}
      </div>
      <Popover placement="bottomLeft" content={uiContent} title="主题设置">
        <SettingOutlined className={styles.headerSetting} style={{ color: gcu.headerStyle.color }} />
      </Popover>
    </Layout.Header>
  );
}

export default React.memo(AppHeader);
