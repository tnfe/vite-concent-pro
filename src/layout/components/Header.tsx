import React from 'react';
import { Layout, Avatar } from 'antd';
import { Blank } from 'components/dumb/general';
import { useC2DefaultMod } from 'services/concent';
import Logo from './Logo';
import SettingIcon from './SettingIcon';
import styles from '../styles/App.module.css';

function AppHeader() {
  const { globalState: gst, globalComputed: gcu } = useC2DefaultMod();

  return (
    <Layout.Header className={styles.header} style={gcu.headerStyle}>
      { !gcu.siderInfo.showSider && <Logo />}
      <div className={styles.headerRightAreaWrap} style={{ color: gcu.headerStyle.color }}>
        <Avatar size={32} src={gst.userIcon} className={styles.userIconWarp} />
        <Blank width="8px" />
        {gst.userName}
        <Blank width="8px" />
        {gcu.settingIconCtrl.showInHeader && <SettingIcon mode="header" />}
      </div>
    </Layout.Header>
  );
}

export default React.memo(AppHeader);
