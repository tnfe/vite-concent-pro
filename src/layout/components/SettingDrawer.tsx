import React from 'react';
import styled from 'styled-components';
import { Row, Col, Drawer, Switch, Tag, Radio, RadioChangeEvent, Slider } from 'antd';
import copy from 'copy-to-clipboard';
import { HeartOutlined, CopyOutlined } from '@ant-design/icons';
import { SketchPicker, ColorResult, SwatchesPicker } from 'react-color';
import { Blank, VerticalBlank } from 'components/dumb/general';
import { useSetupCtx } from 'services/concent';
import { success } from 'services/message';
import { TopHeaderTypes, TopNavBarTypes, SiderViewTypes } from 'configs/constant/sys';
import { CtxDe } from 'types/store';

const MySwatchesPicker: any = styled(SwatchesPicker)`
  width: 450px !important;
`;

const stLoveIcon: React.CSSProperties = { color: 'var(--theme-color)', transform: 'translateY(2px)' };

export function setup(ctx: CtxDe) {
  const { gr, globalState } = ctx;
  const ins = ctx.initState({
    pickerMode: 'fast',
  });

  return {
    state: ins.state,
    closeThemeSettingsDrawer: () => ctx.setGlobalState({ settingDrawerVisible: false }),
    changePickerMode(e: RadioChangeEvent) {
      ins.setState({ pickerMode: e.target.value });
    },
    changeTopHeaderType(e: RadioChangeEvent) {
      gr.setState({ topHeaderType: e.target.value });
    },
    changeTopNavBarType(e: RadioChangeEvent) {
      gr.setState({ topNavBarType: e.target.value });
    },
    changeSiderViewType(e: RadioChangeEvent) {
      gr.changeSiderViewType(e.target.value);
    },
    onWebsiteColorChange(colorResult: ColorResult) {
      gr.changeThemeColor({ themeColor: colorResult.hex, setCustThemeColor: true });
    },
    onSiderThemeChange(checked: boolean) {
      gr.switchSiderTheme(checked);
    },
    changeIsUsingDefaultTheme(checked: boolean) {
      gr.changeIsUsingDefaultTheme(checked);
    },
    onHeaderThemeChange(checked: boolean) {
      gr.switchHeaderTheme(checked);
    },
    onInnerMockChange(checked: boolean) {
      gr.changeIsInnerMock(checked);
    },
    copyColor() {
      copy(globalState.themeColor);
      success(`色值 ${globalState.themeColor} 已复制`, 1);
    },
    changeIsTabPaneHeavyBg: (isTabPaneHeavyBg: boolean) => {
      gr.changeIsTabPaneHeavyBg(isTabPaneHeavyBg);
    },
  };
}

export function SettingDrawer() {
  const { globalState: gst, globalComputed: gcu, settings: se, gr } = useSetupCtx(setup, { tag: 'SettingPanel' });

  return (
    <Drawer title="常用设置" visible={gst.settingDrawerVisible} width="550px"
      onClose={se.closeThemeSettingsDrawer}
    >
      <div style={{ padding: '6px 12px' }}>
        <Tag color="geekblue">站点主题设置：</Tag>
        <Blank />
        <Radio.Group value={se.state.pickerMode} onChange={se.changePickerMode}>
          <Radio value="fast">快速</Radio>
          <Radio value="professional">专业</Radio>
        </Radio.Group>
        <Tag color={gst.themeColor}>{gst.themeColor}</Tag>
        <CopyOutlined className="gHover" onClick={se.copyColor} />
        <VerticalBlank />
        {se.state.pickerMode === 'fast'
          && <MySwatchesPicker color={gst.themeColor} onChange={se.onWebsiteColorChange} />
        }
        {se.state.pickerMode === 'professional'
          && <SketchPicker color={gst.themeColor} onChange={se.onWebsiteColorChange}
            onChangeComplete={se.onWebsiteColorChange}
          />
        }
        <VerticalBlank />
        <Row gutter={[16, { xs: 8, sm: 12, md: 12, lg: 12 }]}>
          <Col span="10"><Tag color="geekblue">恢复默认主题：</Tag></Col>
          <Col span="14">
            <Switch checkedChildren="是" unCheckedChildren="否" checked={gst.isUsingDefaultThemeColor} onChange={se.changeIsUsingDefaultTheme} />
          </Col>
          <Col span="10"><VerticalBlank height="5px" /><Tag color="geekblue">字体颜色透明度：</Tag></Col>
          <Col span="14">
            <Slider value={gst.fontAlpha} min={0} max={100} onChange={gr.changeFontAlpha} />
          </Col>
          <Col span="10"><Tag color="geekblue">导航条项目深色背景：</Tag></Col>
          <Col span="14">
            <Switch checked={gst.isTabPaneHeavyBg} onChange={se.changeIsTabPaneHeavyBg}
              checkedChildren="深" unCheckedChildren="浅"
            />
          </Col>
          <Col span="10"><Tag color="geekblue">暗黑边栏：</Tag></Col>
          <Col span="14">
            <Switch checkedChildren="关闭" unCheckedChildren="开启" checked={gcu.siderThemeSwitchChecked} onChange={se.onSiderThemeChange} />
          </Col>
          <Col span="10"><Tag color="geekblue">暗黑顶栏：</Tag></Col>
          <Col span="14">
            <Switch checkedChildren="关闭" unCheckedChildren="开启" checked={gcu.headerThemeSwitchChecked}
              onChange={se.onHeaderThemeChange}
            />
          </Col>
          <Col span="10"><Tag color="geekblue">顶栏设置：</Tag></Col>
          <Col span="14">
            <Radio.Group value={gst.topHeaderType} onChange={se.changeTopHeaderType}>
              <Radio value={TopHeaderTypes.FIXED}>固定</Radio>
              <Radio value={TopHeaderTypes.FLOWED}>随Y轴滚动</Radio>
              <Radio value={TopHeaderTypes.HIDDEN}><HeartOutlined style={stLoveIcon} />隐藏</Radio>
            </Radio.Group>
          </Col>
          <Col span="10"><Tag color="geekblue">导航条设置：</Tag></Col>
          <Col span="14">
            <Radio.Group value={gst.topNavBarType} onChange={se.changeTopNavBarType}>
              <Radio value={TopNavBarTypes.FIXED}><HeartOutlined style={stLoveIcon} />固定</Radio>
              <Radio value={TopNavBarTypes.FLOWED}>随Y轴滚动</Radio>
              <Radio value={TopNavBarTypes.HIDDEN}>隐藏</Radio>
            </Radio.Group>
          </Col>
          <Col span="10"><Tag color="geekblue">边栏设置：</Tag></Col>
          <Col span="14">
            <Radio.Group value={gst.siderViewType} onChange={se.changeSiderViewType}>
              <Radio value={SiderViewTypes.COLLAPSED}><HeartOutlined style={stLoveIcon} />折叠</Radio>
              <Radio value={SiderViewTypes.NOT_COLLAPSED}>展开</Radio>
              <Radio value={SiderViewTypes.HIDDEN}>隐藏</Radio>
            </Radio.Group>
          </Col>
          <Col span="10"><Tag color="geekblue">innerMock设置：</Tag></Col>
          <Col span="14">
            <Switch checkedChildren="关闭" unCheckedChildren="开启" checked={gst.isInnerMock}
              onChange={se.onInnerMockChange}
            />
          </Col>
        </Row>
      </div>
    </Drawer>
  );
}

export default React.memo(SettingDrawer);
