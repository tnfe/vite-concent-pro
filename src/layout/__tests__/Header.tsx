import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { getRef, SettingsType } from 'concent';
import Header from '../components/Header';
import SettingDrawer, { setup } from '../components/SettingDrawer';
import { CtxM } from 'types/store';

type Ctx = CtxM<{}, '$$default', SettingsType<typeof setup>>;
// 组件实例
let ins = null as unknown as ReactWrapper;
let insPanel = null as unknown as ReactWrapper;
let ref = null as unknown as { ctx: Ctx };

/**
 * @author fancyzhong
 * @priority P0
 * @casetype unit
 */
describe('Header Render', () => {
  beforeAll(() => {
    ins = mount(<Header />);
    insPanel = mount(<SettingDrawer />);

    const insC2Ref = getRef<Ctx>({ moduleName: '$$default', tag: 'SettingPanel' });
    if (!insC2Ref) {
      throw new Error('you may forget add tag for component');
    }
    ref = insC2Ref;
  });

  test('generate Header snapshot', () => {
    expect(toJson(ins)).toMatchSnapshot();
  });

  test('generate SettingPanel snapshot', () => {
    expect(toJson(insPanel)).toMatchSnapshot();
  });

  test('call setup', () => {
    const se = ref.ctx.settings;
    console.log(se);
    expect(se.onHeaderThemeChange({} as any)).toBeFalsy();
    expect(se.onSiderThemeChange({} as any)).toBeFalsy();
    expect(se.onWebsiteColorChange({} as any)).toBeFalsy();
    expect(se.onInnerMockChange({} as any)).toBeFalsy();
    expect(se.closeThemeSettingsDrawer()).toBeFalsy();
    expect(se.changeIsUsingDefaultTheme(true)).toBeFalsy();
  });
});
