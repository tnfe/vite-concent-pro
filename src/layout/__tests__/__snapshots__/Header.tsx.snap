// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Render generate Header snapshot 1`] = `
<Memo(AppHeader)>
  <Header
    className="header"
    style={
      Object {
        "backgroundColor": "white",
        "color": "var(--theme-color)",
        "display": "none",
        "marginLeft": "48px",
        "zIndex": 4,
      }
    }
  >
    <Basic
      className="header"
      prefixCls="ant-layout-header"
      style={
        Object {
          "backgroundColor": "white",
          "color": "var(--theme-color)",
          "display": "none",
          "marginLeft": "48px",
          "zIndex": 4,
        }
      }
      tagName="header"
    >
      <header
        className="ant-layout-header header"
        style={
          Object {
            "backgroundColor": "white",
            "color": "var(--theme-color)",
            "display": "none",
            "marginLeft": "48px",
            "zIndex": 4,
          }
        }
      >
        <div
          className="headerRightAreaWrap"
          style={
            Object {
              "color": "var(--theme-color)",
            }
          }
        >
          <Avatar
            className="userIconWarp"
            shape="circle"
            size={32}
            src=""
          >
            <span
              className="ant-avatar ant-avatar-circle userIconWarp"
              style={
                Object {
                  "fontSize": 18,
                  "height": 32,
                  "lineHeight": "32px",
                  "width": 32,
                }
              }
            >
              <img
                onError={[Function]}
                src=""
              />
            </span>
          </Avatar>
          <Blank
            width="8px"
          >
            <div
              style={
                Object {
                  "display": "inline-block",
                  "height": "16px",
                  "width": "8px",
                }
              }
            />
          </Blank>
          <Blank
            width="8px"
          >
            <div
              style={
                Object {
                  "display": "inline-block",
                  "height": "16px",
                  "width": "8px",
                }
              }
            />
          </Blank>
        </div>
      </header>
    </Basic>
  </Header>
</Memo(AppHeader)>
`;

exports[`Header Render generate SettingPanel snapshot 1`] = `
<Memo(SettingDrawer)>
  <DrawerWrapper
    onClose={[Function]}
    title="常用设置"
    visible={false}
    width="550px"
  >
    <Drawer
      closable={true}
      height={256}
      keyboard={true}
      level={null}
      mask={true}
      maskClosable={true}
      onClose={[Function]}
      placement="right"
      prefixCls="ant-drawer"
      push={
        Object {
          "distance": 180,
        }
      }
      title="常用设置"
      visible={false}
      width="550px"
    >
      <DrawerWrapper
        afterVisibleChange={[Function]}
        className=""
        defaultOpen={false}
        duration=".3s"
        ease="cubic-bezier(0.78, 0.14, 0.15, 0.86)"
        forceRender={false}
        getContainer="body"
        handler={false}
        keyboard={true}
        level={null}
        maskClosable={true}
        maskStyle={Object {}}
        onChange={[Function]}
        onClose={[Function]}
        open={false}
        placement="right"
        prefixCls="ant-drawer"
        showMask={true}
        style={
          Object {
            "transform": undefined,
            "zIndex": undefined,
          }
        }
        width="550px"
        wrapperClassName=""
      >
        <PortalWrapper
          forceRender={false}
          getContainer="body"
          visible={false}
          wrapperClassName=""
        />
      </DrawerWrapper>
    </Drawer>
  </DrawerWrapper>
</Memo(SettingDrawer)>
`;
